Tests/UnitTests/Core/Utilities/DelegateTests.cpp

```diff
@@ -199,4 +201,212 @@ BOOST_AUTO_TEST_CASE(StatefullLambdas) {
+      BOOST_CHECK_EQUAL(destructorCalled, false);
+      // This doesn't compile: owning delegate cannot accept raw pointer
+      // instance
+      // d.connect<&CheckDestructor::func>(s.get());
+      d.connect<&CheckDestructor::func>(std::move(s));
+      BOOST_CHECK_EQUAL(destructorCalled, false);
+      BOOST_CHECK_EQUAL(d(), 4);
+      BOOST_CHECK_EQUAL(destructorCalled, false);
+      d.disconnect();
+      BOOST_CHECK_EQUAL(destructorCalled, true);
+    }
+    // destructor called after owning delegate goes out of scope
+    BOOST_CHECK_EQUAL(destructorCalled, true);
+  }
+
+  {
+    OwningDelegate<int()> d;
+    d.connect<&owningTest>();
+    BOOST_CHECK_EQUAL(d(), 8);
+
+    d.disconnect();
+    d.connect<&owningTest>();
+    BOOST_CHECK_EQUAL(d(), 8);
+
+    d.disconnect();
+    d.connect(owningTest2);
+    BOOST_CHECK_EQUAL(d(), 8);
+  }
+}
+
+struct DelegateInterface {
+  virtual ~DelegateInterface() = 0;
+
+  virtual std::string func() const { return "base"; }
+};
+inline DelegateInterface::~DelegateInterface() = default;
+
+struct ConcreteDelegate : public DelegateInterface {
+  std::string func() const final { return "derived"; }
+};
+
+struct SeparateDelegate {
+  std::string func() const { return "separate"; }
+};
+
+BOOST_AUTO_TEST_CASE(NonVoidDelegateTest) {
+  // check void behavior with virtuals
+  {
+    Delegate<std::string(), void> d;
+    ConcreteDelegate c;
+    d.connect<&ConcreteDelegate::func>(&c);
+    BOOST_CHECK_EQUAL(d(), "derived");
+
+    // does not compile: delegate won't hand out void pointer
+    // d.instance();
+  }
+  {
+    Delegate<std::string(), void> d;
+    ConcreteDelegate c;
+    d.connect<&DelegateInterface::func>(&c);
+    BOOST_CHECK_EQUAL(
+        d(), "derived");  // <- even if you plug in the base class member
+                          // pointer you get the derived class call
+  }
+
+  {
+    Delegate<std::string(), DelegateInterface> d;
+    ConcreteDelegate c;
+    d.connect<&ConcreteDelegate::func>(&c);
+    BOOST_CHECK_EQUAL(d(), "derived");
+
+    const auto* instance = d.instance();
+    static_assert(
+        std::is_same_v<
+            std::remove_const_t<std::remove_pointer_t<decltype(instance)>>,
+            DelegateInterface>,
+        "Did not get correct instance pointer");
+    BOOST_CHECK_NE(dynamic_cast<const DelegateInterface*>(instance), nullptr);
+    BOOST_CHECK_NE(dynamic_cast<const ConcreteDelegate*>(instance), nullptr);
+  }
+  {
+    Delegate<std::string(), ConcreteDelegate> d;
+    ConcreteDelegate c;
+    d.connect<&ConcreteDelegate::func>(&c);
+    BOOST_CHECK_EQUAL(d(), "derived");
+
+    const auto* instance = d.instance();
+    static_assert(
+        std::is_same_v<
+            std::remove_const_t<std::remove_pointer_t<decltype(instance)>>,
+            ConcreteDelegate>,
+        "Did not get correct instance pointer");
+    BOOST_CHECK_NE(dynamic_cast<const DelegateInterface*>(instance), nullptr);
```

Comment:
That's perfect, actually - I will be able to define dedicated interfaces for the navigation delegates.